rest_api/sawtooth_rest_api/routes.py

```diff
@@ -178,56 +139,60 @@ def _try_validator_request(self, message_type, content):
         try:
             response = future.result(timeout=timeout)
         except FutureTimeoutError:
-            raise web.HTTPServiceUnavailable(reason=timeout_msg)
+            raise errors.ValidatorUnavailable()
 
         return response.content
 
     @staticmethod
-    def _try_response_parse(proto, response):
+    def _try_response_parse(proto, response, traps=None):
         """
         Parses a protobuf response from the validator
         Raises common validator error statuses as HTTP errors
         """
-        unknown_msg = 'An unknown error occured with your request'
-        notfound_msg = 'There is no resource at that root, address or prefix'
-
         parsed = proto()
         parsed.ParseFromString(response)
+        traps = traps or []
 
+        # Not every protobuf has every status enum, so pass AttributeErrors
+        try:
+            traps.append(errors.Unknown(proto.ERROR))
+        except AttributeError:
+            pass
         try:
-            if parsed.status == proto.ERROR:
-                raise web.HTTPInternalServerError(reason=unknown_msg)
-            if parsed.status == proto.NORESOURCE:
-                raise web.HTTPNotFound(reason=notfound_msg)
+            traps.append(errors.MissingGenesis(proto.NOGENESIS))
         except AttributeError:
-            # Not every protobuf has every status, so pass AttributeErrors
             pass
 
+        for trap in traps:
+            trap.check(parsed.status)
+
         return MessageToDict(parsed, preserving_proto_field_name=True)
 
     @staticmethod
-    def _wrap_response(data=None, head=None, link=None):
+    def _safe_get(obj, key, default=''):
+        """
+        aiohttp very helpfully parses param strings to replace '+' with ' '
+        This is very bad when your block ids contain meaningful +'s
+        """
+        return obj.get(key, default).replace(' ', '+')
```

Comment:
The problem is, by the time the string gets to the route handlers, it has had `'+'` swapped out for `' '`. You can't b64decode it by then, because the +'s are already gone. Currently, there are two properties that this is happening for, and they each have their own more correct solution as I see it.

`request.url.query`:
This covers query string parameters. It is possible to get the raw query string (i.e. `'head=n40+2/3h...'`), which should solve the issue, _but_ that does mean parsing the query string manually.

`request.match_info`:
This is where parameters in the path come from, and don't really have a raw equivalent. You can get the whole url string, but unlike with query parameters, there is not an immediately obvious programatic way to no which URL segment is the parameter you are looking for. Probably what we'd have to do is write a custom middleware to build a custom `match_info` equivalent.

I think the _best_ and _easiest_ solution is just to have block ids that are url-safe. We could maybe use `urlsafe_b64encode_ somewhere in the validator to do that, but I understand from Dan, that the ids are changing soon anyway. Might be best to punt on this for now.